[Shawn Lim] iP

lwlshawn opened the pull request:
# Frosty

> "Yo-ho-ho, Merry Christmas!" - No one ever, [probably](https://en.wikipedia.org/wiki/Santa_Claus).

Frosty is a ~~(eventually or maybe never)~~ christmas themed personal assistant! It can help you:

* Keep a list of tasks and save them for future use
* Mark tasks as done, attach dates to some tasks
* Be filled with christmas spirit :slightly_smiling_face:

It supports the following commands:
1. list - displays a list of your current tasks
2. todo <description> - Adds a "todo" task with a non-empty description
3. deadline <description> /by <date> - Adds a "deadline" task with a non-empty description and date. Date should be given in the form yyyy-mm-dd (e.g. 2020-02-20)
4. event <description> /at <date> - Adds  an "event" task. Same requirements as deadline tasks
5. done <number> - Marks a task in your list as complete
6. delete <number> - Removes a task from your list

Features:
- [x] Todo tasks
- [x]  Deadline tasks
- [x]  Event tasks
- [ ]  Christmas spirit (coming soon<sup>tm</sup>)

For reasons, enjoy the following snippet of code taken from `main`:
```javascript
    public static void main(String[] args) {
        Application.launch(Main.class, args);
    }
```

Commit:
Add Gradle support

lwlshawn committed:
Level-1: Greet, Echo, Exit complete

lwlshawn committed:
Implemented level 2

lwlshawn committed:
Added Task.java, edited some UI elements of Duke.java, implemented level 3

lwlshawn committed:
Added Deadline Event Todo that extend Task. Completed Level 4 functionality

lwlshawn committed:
Automated testing doesn't work, script is broken and requires manual calling of FC with direct address

lwlshawn committed:
Added basic error handling

lwlshawn committed:
Level-6 completed. Outstanding issues for future weeks; error checking is currently very basic and automated script does not work.

lwlshawn committed:
Error checking is now more comprehensive, updated the input and expected text for automated checking to now do a basic check covering Todo, Deadline, Event tasks, as well as done and delete functionality

lwlshawn committed:
Refactored out direct ArrayList management to TaskList class.

lwlshawn committed:
Save Load functionality, added some minor functionalities to the Todo Event Deadline classes, made the Task class abstract.

lwlshawn committed:
Minimal date functionality for event and deadline. Can understand format yyyy-mm-dd and prints in MMM dd yyyy.

lwlshawn committed:
Merge branch 'branch-Level-7'

lwlshawn committed:
Merge branch 'branch-Level-8'

lwlshawn committed:
Created Ui Class with basic functionality. Some aspects of Ui handling not yet ported over completely.

lwlshawn committed:
Refactored Ui slightly to make seperation of concerns more clean. Parser class initialised but still blank at this point

lwlshawn committed:
Parser and command logic drafted. Not yet implemented; this commit is before the significant refactoring of Duke.java

lwlshawn committed:
Adding delete and event commands are currently broken. I want to shift date understanding to the parser instead of in the delete/event classes in any case. Save load also not implemented yet, but basic add, mark done, delete, list commands are up and running

lwlshawn committed:
Removed a lot of code that is no longer relevant, add deadline/event are functional again. Last outstanding part of refactor is to make save/load work with new structure.

lwlshawn committed:
Save and load functionality is back. Main remaining issues before OOP refactor is complete is a few more print statements to be handed to ui, redoing how exceptions are handled, and adding a few more exception checks (delete IOOB not caught right now)

lwlshawn committed:
All exception handling shifted to Main, all ui duties now correctly with Ui class, basic functionalities up to level 8 are all in place.

lwlshawn committed:
divided classes into packages

lwlshawn committed:
Found and fixed print formatting bugs. Implemented junit tests for AddCommand, one test for each existing type of task.

lwlshawn committed:
More unit tests added, checking some basic aspects of Parser functionality

lwlshawn committed:
Added displayline functionality to UI after seeing how the jar file looks in a terminal. Tested jar and it appears to satisfy basic functionality.

lwlshawn committed:
Added find functionality

FindCommand class added, containing the logic for the new function.
Parser class changed to recognize the find command and added method
to Ui class for displaying the found items.

lwlshawn committed:
Added javadocs to the command package

lwlshawn committed:
Added javadocs to the task package

lwlshawn committed:
Javadocs for remaining classes added

Documents for public methods in ui and tasklist that are very
self-explanatory have been omitted.

lwlshawn committed:
Updated code to improve style

Changed naming of test methods, edited use of white space to match
given style guide.

lwlshawn committed:
Merge branch 'branch-A-CodingStandard'

lwlshawn committed:
Merge branch 'branch-A-JavaDoc'

# Conflicts:
#	src/main/java/duke/Storage.java
#	src/main/java/duke/command/DeleteCommand.java

lwlshawn committed:
Merge branch 'branch-Level-9'

lwlshawn committed:
Add Launcher.java and fixed some bugs

After all the chain of commits, some places invoked the Storage
constructor without the correct number of arguments. Deadline and
Event were also bugged due to not calling parent constructor.

Bugs fixed and program passes test suite again. Basic UI display
is now functional.

lwlshawn committed:
Complete JavaFX part 2

lwlshawn committed:
Part3 Iteration 1

lwlshawn committed:
Javafx Part 3 Iteration 2

lwlshawn committed:
Javafx Part 3 Iteration 3

lwlshawn committed:
Complete Javafx tutorial

lwlshawn committed:
Refactor app to work with GUI

Main class has switched following the JavaFX tutorial. Requires
refactoring of UI class that handled printing to screen and some
rewriting of command execute logic.

UI object now handles generating the correct strings, and returns
them to the GUI half to display instead of printing it directly.

Done this way since the UI class was already responsible for
correctly formatting messages, allowed sensible reuse of existing
codebase.

Some functionality left to be done; load functionality and exit
functionality is not properly present. Some Ui commands are also
not fully operational and display placeholder messages.

lwlshawn committed:
Complete basic GUI implementation

All basic features are now operational and display correctly.
Remaining significant issue is that text in GUI display does not
resize automatically and only shows a tiny bit of text.

lwlshawn committed:
Merge branch 'add-gradle-support' of https://github.com/nus-cs2103-AY2122S1/ip into branch-A-Gradle

lwlshawn committed:
Gradle run implemented

Gradle build will be left until checkstyle and other things are
complete.

lwlshawn committed:
Tidy codebase to match checkstyle

lwlshawn committed:
Implement gradle build and remove checkstyle errors

All remaining checkstyle errors fixed, gradle build successful

huizhuansam commented on src/main/java/duke/Parser.java:
```diff
@@ -0,0 +1,77 @@
+    }
+
+    /**
+     * Takes in a String representing user input, and parses it, producing the appropriate Command.
```
Do you think it will be neater if a line is left between the description and parameter section of the Javadoc? The same comment applies across the other files in the project.
Reason: Javadoc convention

huizhuansam commented on src/main/java/duke/Storage.java:
```diff
@@ -0,0 +1,110 @@
+     * Currently filename represents where the data is to be saved, and it is hard-coded in Duke.java.
+     * @param filename is the file the application will write to, in the data folder.
+     */
+    public Storage (String filename, Ui ui) {
```
Might just be me nitpicking, but camelCase `fileName` might be preferable over `filename`; thoughts?
Reason: parameter naming guidelines

huizhuansam commented on src/main/java/duke/Storage.java:
```diff
@@ -0,0 +1,110 @@
+     * @param tasklist contains the tasks from the current session.
+     * @throws IOException if writing to the text file fails, or initialising the writer fails.
+     */
+    public void save(TaskList tasklist) throws IOException {
```
Perhaps camelCase `taskList` can be used here.
Reason: naming conventions

huizhuansam commented on src/main/java/duke/Storage.java:
```diff
@@ -0,0 +1,110 @@
+    }
+
+    //converts a Task into a string to be saved to a txt file.
+    private String formatForSave(Task e) {
```
Perhaps a more descriptive name for the method can be used because `formatForSave` might cause other devs to assume there's a "Save" subroutine inside.

huizhuansam replied:
Just another nitpick, parameter `e` could be given a better name, because further down the code where you are calling `e.getType()`, `e.getFlag()`, devs might wonder what is `e` and have to trace back to the method signature.

huizhuansam commented on src/main/java/duke/Storage.java:
```diff
@@ -0,0 +1,110 @@
+    }
+
+    //dual of formatForSave. Converts from saved string back to Task.
+    private Task formatForLoad(String s) throws DukeException {
```
See comments on line 80.

huizhuansam commented on src/main/java/duke/TaskList.java:
```diff
@@ -0,0 +1,41 @@
+ * Java's arraylist with a more limited interface, implemented due to course requirements.
+ */
+public class TaskList {
+    private ArrayList<Task> tl;
```
Hmm, will it be better if this variable is named differently? `tl` might lose its meaning once the codebase grows large...

huizhuansam commented on src/main/java/duke/command/AddCommand.java:
```diff
@@ -0,0 +1,60 @@
+ * Implements the logic for adding a Task to the users list.
+ */
+public class AddCommand extends Command {
+    private final String type;
```
Do you want to explain what `type` means here?

huizhuansam commented on src/main/java/duke/command/Command.java:
```diff
@@ -0,0 +1,23 @@
+ * execute to achieve its specific task. Common features to all Commands implemented here.
+ */
+public abstract class Command {
+    protected boolean isDone = false;
```
I'm slightly confused here, care to explain what is `isDone`?

huizhuansam commented on src/main/java/duke/task/Task.java:
```diff
@@ -0,0 +1,68 @@
+
+public abstract class Task {
+    protected String label;
+    private boolean flag;
```
Hmm, `flag` does seem to have ambiguous meaning, don't you think?

huizhuansam commented on src/main/java/duke/task/Todo.java:
```diff
@@ -0,0 +1,34 @@
+        return "[T]" + super.toString();
+    }
+
+    @Override
```
Do you want to explain why `getDate()` returns an empty string? Might be useful for other devs.

huizhuansam reviewed (commented):
Hi Shawn, I think your code has some minor issues in terms of readability. Otherwise, I think this is a pretty solid implementation, based on the general structure of the program. I also liked the detailed Javadoc descriptions, especially on file headers. I left some comments here, feel free to reply to them if you need further clarification.

Best, 
Hui Zhuan

noobmaster19 commented on src/main/java/duke/task/Todo.java:
```diff
@@ -0,0 +1,34 @@
+    public String toString() {
+        return "[T]" + super.toString();
+    }
+
```
Hey, same thing here, might want to explain why it returns an empty string :)

noobmaster19 commented on src/main/java/duke/Parser.java:
```diff
@@ -0,0 +1,77 @@
+            String[] temp = in.split(" ");
+            return new DeleteCommand(Integer.parseInt(temp[1]) - 1);
+        } else if (in.equals("save")) {
+            return new SaveCommand();
```
Nearly organized!

noobmaster19 commented on src/main/java/duke/MainWindow.java:
```diff
@@ -0,0 +1,59 @@
+    }
+
+    public void setDuke(Duke d) {
+        duke = d;
```
maybe want to take in the parameter as duke and d the assignemnt as this.duke = duke

noobmaster19 commented on src/main/java/duke/Main.java:
```diff
@@ -0,0 +1,31 @@
+    private Duke duke = new Duke();
+
+    @Override
+    public void start(Stage stage) {
```
Great organanization for the UI

lwlshawn committed:
Refactor Ui class

A number of methods became deprecated after unnecessarily
implementing javaFX.

Deprecated methods that are no longer needed have been removed.

Done to improve code quality.

Removing javaFX will improve the experience for everyone.

lwlshawn committed:
AddCommand: add assert to command flow

If else chain used an else clause, instead of explicitly checking
that the type was "event". This could result in problems if the
type was not one of the three recognised types.

The assumption is now explicit. The code checks sequentially
for each of the three accepted types, and uses an assertion
where the code should not execute.

This checks for a serious error in the execution, since it means a
command was incorrectly parsed as an AddCommand when the syntax was
incorrect.

lwlshawn committed:
Refactor Parser class

Readability of parse method was very poor with many nested if
else clauses and poor application of SLAP.

Code has been refactored to improve readability and if else clauses
changed to switch clauses.

Done to improve readability, simplify future additions and
make maintainance easier.

Refactoring broke invalidcommand, deadline and event logic. To
be fixed next.

lwlshawn committed:
Fix bugs from refactor and associated bugs uncovered

Refactor of parse method broke some functions of the application.

Issue was caused by split removing spaces, and join not re-adding
the spaces. In addition, realised that exceptions thrown were not
properly printing error messages to the user.

All the above listed issues were fixed. Errors are now printed
to the user correctly, and when using join, removed spaces are
re-included.

lwlshawn committed:
Merge pull request #1 from lwlshawn/branch-A-CodeQuality

Branch a code quality

lwlshawn committed:
Merge branch 'master' of https://github.com/lwlshawn/ip

lwlshawn committed:
Merge branch 'master' into branch-A-Assertions

lwlshawn committed:
Merge pull request #2 from lwlshawn/branch-A-Assertions

Branch a assertions

lwlshawn committed:
Merge branch 'master' of https://github.com/lwlshawn/ip

lwlshawn committed:
Remove debug print statements

Leftover code from fixing bugs was removed.

lwlshawn committed:
Implement basic update command

It is now possible to update the description of existing items
in the tasklist.

Error handling and implementation of updating dates is next.

lwlshawn committed:
Implement C-Update

Implementation of update functionality for BCD-extension
requirement.

Date is now a universal feature of Tasks. Asserts have been
added to Todo class to ensure date related methods are not wrongly
called on Todo Tasks in the program.

Done this way since more Tasks (Deadline, Event) have dates then
tasks that do not (Todo).

Uncovered bugs in parser error reporting logic. Have documented
them in comments to be solved at a future date.

noobmaster19 reviewed (commented):
Here

lwlshawn committed:
Correct style issues and junit tests

Fixed identified style issues from iP code quality feedback. Tests
were broken from implementation of gradle, issue has been resolved.

lwlshawn committed:
Fix style issues in JUnit tests

lwlshawn committed:
Fix header comment style issue

Description and params were not seperated by a line. Fixed in all
files.

lwlshawn committed:
Fix issues outlined in PR review

All relevant issues outlined by peers in PR have been addressed

lwlshawn committed:
Remove DoneCommand

Code is no longer relevant after GUI implemented. "done" also
removed from parser logic.

lwlshawn committed:
Fix bugs in updatecommand and undo remove of done command

Updatecommand did not recognise when user incorrectly tried to
update date of todo tasks; this has been remedied. Done command
was mistakenly removed; this is to mark tasks as done, and not
to exit the application. Change has been reverted

lwlshawn committed:
Remove InvalidCommand class

Instead of using a command, now incorrect instructions simply
result in a DukeException thrown with a notice. Done to standardise
how invalid input errors were handled from the user.

lwlshawn committed:
Remove old images

New images were added to flavor the application in earlier commit.
Old images no longer used now removed from repo.

lwlshawn committed:
Set theme jekyll-theme-architect

lwlshawn committed:
Update Readme

lwlshawn committed:
Add detail to README

lwlshawn committed:
Fix test

Updated test code to match new expected behavior

lwlshawn committed:
Delete README.md
